File: radar-commons-android/src/main/java/org/radarcns/android/device/DeviceService.java
/*
 * Copyright 2017 The Hyve
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.radarcns.android.device;

import android.app.Service;
import android.bluetooth.BluetoothAdapter;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.os.Binder;
import android.os.Bundle;
import android.os.Handler;
import android.os.IBinder;
import android.os.Parcel;
import android.os.ResultReceiver;
import android.support.annotation.CallSuper;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.content.LocalBroadcastManager;
import android.util.Pair;

import org.radarcns.android.RadarApplication;
import org.radarcns.android.RadarConfiguration;
import org.radarcns.android.auth.AppAuthState;
import org.radarcns.android.auth.AppSource;
import org.radarcns.android.auth.portal.ManagementPortalService;
import org.radarcns.android.data.ReadableDataCache;
import org.radarcns.android.data.TableDataHandler;
import org.radarcns.android.kafka.ServerStatusListener;
import org.radarcns.android.util.BundleSerialization;
import org.radarcns.data.RecordData;
import org.radarcns.kafka.ObservationKey;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.util.Collections;
import java.util.Map;
import java.util.Objects;
import java.util.Set;
import java.util.concurrent.ThreadLocalRandom;

import static org.radarcns.android.RadarConfiguration.SOURCE_ID_KEY;
import static org.radarcns.android.auth.portal.ManagementPortalService.MANAGEMENT_PORTAL_REGISTRATION;
import static org.radarcns.android.device.DeviceServiceProvider.NEEDS_BLUETOOTH_KEY;
import static org.radarcns.android.device.DeviceServiceProvider.SOURCE_KEY;
import static org.radarcns.android.kafka.ServerStatusListener.Status.UNAUTHORIZED;

/**
 * A service that manages a DeviceManager and a TableDataHandler to send addToPreferences the data of a
 * wearable device and send it to a Kafka REST proxy.
 *
 * Specific wearables should extend this class.
 */
@SuppressWarnings("WeakerAccess")
public abstract class DeviceService<T extends BaseDeviceState> extends Service implements DeviceStatusListener {
    private static final String PREFIX = "org.radarcns.android.";
    public static final String SERVER_STATUS_CHANGED = PREFIX + "ServerStatusListener.Status";
    public static final String SERVER_RECORDS_SENT_TOPIC = PREFIX + "ServerStatusListener.topic";
    public static final String SERVER_RECORDS_SENT_NUMBER = PREFIX + "ServerStatusListener.lastNumberOfRecordsSent";
    public static final String CACHE_TOPIC = PREFIX + "DataCache.topic";
    public static final String CACHE_RECORDS_UNSENT_NUMBER = PREFIX + "DataCache.numberOfRecords.first";
    public static final String CACHE_RECORDS_SENT_NUMBER = PREFIX + "DataCache.numberOfRecords.second";
    public static final String DEVICE_SERVICE_CLASS = PREFIX + "DeviceService.getClass";
    public static final String DEVICE_STATUS_CHANGED = PREFIX + "DeviceStatusListener.Status";
    public static final String DEVICE_STATUS_NAME = PREFIX + "DeviceManager.getName";
    public static final String DEVICE_CONNECT_FAILED = PREFIX + "DeviceStatusListener.deviceFailedToConnect";

    private final ObservationKey key = new ObservationKey();

    /** Stops the device when bluetooth is disabled. */
    private final BroadcastReceiver mBluetoothReceiver = new BroadcastReceiver() {
        @Override
        public void onReceive(Context context, Intent intent) {
            final String action = intent.getAction();

            if (Objects.equals(action, BluetoothAdapter.ACTION_STATE_CHANGED)) {
                final int state = intent.getIntExtra(
                        BluetoothAdapter.EXTRA_STATE, BluetoothAdapter.ERROR);
                switch (state) {
                    case BluetoothAdapter.STATE_TURNING_OFF: case BluetoothAdapter.STATE_OFF:
                        logger.warn("Bluetooth is off");
                        stopDeviceManager(unsetDeviceManager());
                        break;
                    default:
                        logger.debug("Bluetooth is in state {}", state);
                        break;
                }
            }
        }
    };

    private static final Logger logger = LoggerFactory.getLogger(DeviceService.class);
    private TableDataHandler dataHandler;
    private DeviceManager<T> deviceScanner;
    private DeviceBinder mBinder;
    private boolean hasBluetoothPermission;
    private AppSource source;

    @CallSuper
    @Override
    public void onCreate() {
        logger.info("Creating DeviceService {}", this);
        super.onCreate();
        mBinder = createBinder();

        if (isBluetoothConnectionRequired()) {
            IntentFilter intentFilter = new IntentFilter(BluetoothAdapter.ACTION_STATE_CHANGED);
            registerReceiver(mBluetoothReceiver, intentFilter);
        }

        synchronized (this) {
            deviceScanner = null;
        }
    }

    @CallSuper
    @Override
    public void onDestroy() {
        logger.info("Destroying DeviceService {}", this);
        super.onDestroy();

        if (isBluetoothConnectionRequired()) {
            // Unregister broadcast listeners
            unregisterReceiver(mBluetoothReceiver);
        }
        stopDeviceManager(unsetDeviceManager());
        ((RadarApplication)getApplicationContext()).onDeviceServiceDestroy(this);

        if (dataHandler != null) {
            try {
                dataHandler.close();
            } catch (IOException e) {
                // do nothing
            }
        }
    }

    @Nullable
    @Override
    public IBinder onBind(Intent intent) {
        doBind(intent, true);
        return mBinder;
    }

    @CallSuper
    @Override
    public void onRebind(Intent intent) {
        doBind(intent, false);
    }

    private void doBind(Intent intent, boolean firstBind) {
        logger.info("Received (re)bind in {}", this);
        Bundle extras = BundleSerialization.getPersistentExtras(intent, this);
        onInvocation(extras);

        RadarApplication application = (RadarApplication)getApplicationContext();
        application.onDeviceServiceInvocation(this, extras, firstBind);
    }

    @Override
    public boolean onUnbind(Intent intent) {
        logger.info("Received unbind in {}", this);
        return true;
    }

    @Override
    public void deviceFailedToConnect(String deviceName) {
        Intent statusChanged = new Intent(DEVICE_CONNECT_FAILED);
        statusChanged.putExtra(DEVICE_SERVICE_CLASS, getClass().getName());
        statusChanged.putExtra(DEVICE_STATUS_NAME, deviceName);
        LocalBroadcastManager.getInstance(this).sendBroadcast(statusChanged);
    }


    private void broadcastDeviceStatus(String name, DeviceStatusListener.Status status) {
        Intent statusChanged = new Intent(DEVICE_STATUS_CHANGED);
        statusChanged.putExtra(DEVICE_STATUS_CHANGED, status.ordinal());
        statusChanged.putExtra(DEVICE_SERVICE_CLASS, getClass().getName());
        if (name != null) {
            statusChanged.putExtra(DEVICE_STATUS_NAME, name);
        }
        LocalBroadcastManager.getInstance(this).sendBroadcast(statusChanged);
    }

    private void broadcastServerStatus(ServerStatusListener.Status status) {
        Intent statusIntent = new Intent(SERVER_STATUS_CHANGED);
        statusIntent.putExtra(SERVER_STATUS_CHANGED, status.ordinal());
        statusIntent.putExtra(DEVICE_SERVICE_CLASS, getClass().getName());
        LocalBroadcastManager.getInstance(this).sendBroadcast(statusIntent);
    }

    @Override
    public void deviceStatusUpdated(DeviceManager deviceManager, DeviceStatusListener.Status status) {
        switch (status) {
            case DISCONNECTED:
                stopDeviceManager(deviceManager);
                synchronized (this) {
                    deviceScanner = null;
                }
                break;
            default:
                // do nothing
                break;
        }
        broadcastDeviceStatus(deviceManager.getName(), status);
    }

    private synchronized DeviceManager unsetDeviceManager() {
        DeviceManager tmpManager = deviceScanner;
        deviceScanner = null;
        return tmpManager;
    }

    private void stopDeviceManager(DeviceManager deviceManager) {
        if (deviceManager != null) {
            if (!deviceManager.isClosed()) {
                try {
                    deviceManager.close();
                } catch (IOException e) {
                    logger.warn("Failed to close device scanner", e);
                }
            }
        }
    }

    /**
     * New device manager for the current device.
     */
    protected abstract DeviceManager<T> createDeviceManager();

    protected T getState() {
        DeviceManager<T> localManager = getDeviceManager();
        if (localManager != null) {
            return localManager.getState();
        }

        T state = getDefaultState();
        ObservationKey stateKey = state.getId();
        stateKey.setProjectId(key.getProjectId());
        stateKey.setUserId(key.getUserId());
        stateKey.setSourceId(key.getSourceId());
        return state;
    }

    /**
     * Default state when no device manager is active.
     */
    protected abstract T getDefaultState();

    public T startRecording(@NonNull Set<String> acceptableIds) {
        DeviceManager localManager = getDeviceManager();
        if (key.getUserId() == null) {
            throw new IllegalStateException("Cannot start recording: user ID is not set.");
        }
        if (localManager == null) {
            if (isBluetoothConnectionRequired() && BluetoothAdapter.getDefaultAdapter() != null && !BluetoothAdapter.getDefaultAdapter().isEnabled()) {
                throw new IllegalStateException("Cannot start recording without Bluetooth");
            }
            logger.info("Starting recording");
            synchronized (this) {
                if (deviceScanner == null) {
                    deviceScanner = createDeviceManager();
                    deviceScanner.start(acceptableIds);
                }
            }
        }
        return getDeviceManager().getState();
    }

    public void stopRecording() {
        stopDeviceManager(unsetDeviceManager());
        logger.info("Stopped recording {}", this);
    }

    protected class DeviceBinder extends Binder implements DeviceServiceBinder {
        @SuppressWarnings("unchecked")
        @Nullable
        @Override
        public RecordData<Object, Object> getRecords(
                @NonNull String topic, int limit) throws IOException {
            TableDataHandler localDataHandler = getDataHandler();
            if (localDataHandler == null) {
                return null;
            }
            return localDataHandler.getCache(topic).getRecords(limit);
        }

        @Override
        public T getDeviceStatus() {
            return getState();
        }

        @Override
        public String getDeviceName() {
            DeviceManager localManager = getDeviceManager();
            if (localManager == null) {
                return null;
            }
            return localManager.getName();
        }

        @Override
        public T startRecording(@NonNull Set<String> acceptableIds) {
            return DeviceService.this.startRecording(acceptableIds);
        }

        @Override
        public void stopRecording() {
            DeviceService.this.stopRecording();
        }

        @Override
        public ServerStatusListener.Status getServerStatus() {
            TableDataHandler localDataHandler = getDataHandler();
            if (localDataHandler == null) {
                return ServerStatusListener.Status.DISCONNECTED;
            }
            return localDataHandler.getStatus();
        }

        @Override
        public Map<String,Integer> getServerRecordsSent() {
            TableDataHandler localDataHandler = getDataHandler();
            if (localDataHandler == null) {
                return Collections.emptyMap();
            }
            return localDataHandler.getRecordsSent();
        }

        @Override
        public void updateConfiguration(Bundle bundle) {
            onInvocation(bundle);
        }

        @Override
        public Pair<Long, Long> numberOfRecords() {
            long unsent = -1L;
            long sent = -1L;
            TableDataHandler localDataHandler = getDataHandler();
            if (localDataHandler != null) {
                for (ReadableDataCache cache : localDataHandler.getCaches()) {
                    Pair<Long, Long> pair = cache.numberOfRecords();
                    if (pair.first != -1L) {
                        if (unsent == -1L) {
                            unsent = pair.first;
                        } else {
                            unsent += pair.first;
                        }
                    }
                    if (pair.second != -1L) {
                        if (sent == -1L) {
                            sent = pair.second;
                        } else {
                            sent += pair.second;
                        }
                    }
                }
            }
            return new Pair<>(unsent, sent);
        }

        @Override
        public boolean needsBluetooth() {
            return isBluetoothConnectionRequired();
        }

        @Override
        public void setDataHandler(TableDataHandler dataHandler) {
            DeviceService.this.setDataHandler(dataHandler);
        }

        @Override
        public boolean onTransact(int code, Parcel data, Parcel reply, int flags) {
            throw new UnsupportedOperationException();
        }
    }

    /**
     * Override this function to get any parameters from the given intent.
     * Bundle classloader needs to be set correctly for this to work.
     *
     * @param bundle intent extras that the activity provided.
     */
    @CallSuper
    protected void onInvocation(@NonNull Bundle bundle) {
        source = bundle.getParcelable(SOURCE_KEY);
        if (source == null) {
            source = new AppSource(-1L, null, null, null, true);
        }
        if (source.getSourceId() != null) {
            key.setSourceId(source.getSourceId());
        }
        AppAuthState authState = AppAuthState.Builder.from(bundle).build();
        key.setProjectId(authState.getProjectId());
        key.setUserId(authState.getUserId());

        setHasBluetoothPermission(bundle.getBoolean(NEEDS_BLUETOOTH_KEY, false));
    }

    public synchronized TableDataHandler getDataHandler() {
        return dataHandler;
    }

    public synchronized void setDataHandler(TableDataHandler dataHandler) {
        this.dataHandler = dataHandler;
    }

    public synchronized DeviceManager<T> getDeviceManager() {
        return deviceScanner;
    }

    /** Get the service local binder. */
    @NonNull
    protected DeviceBinder createBinder() {
        return new DeviceBinder();
    }

    protected void setHasBluetoothPermission(boolean isRequired) {
        boolean oldBluetoothNeeded = isBluetoothConnectionRequired();
        hasBluetoothPermission = isRequired;
        boolean newBluetoothNeeded = isBluetoothConnectionRequired();

        if (oldBluetoothNeeded && !newBluetoothNeeded) {
            unregisterReceiver(mBluetoothReceiver);
        } else if (!oldBluetoothNeeded && newBluetoothNeeded) {
            // Register for broadcasts on BluetoothAdapter state change
            IntentFilter filter = new IntentFilter(BluetoothAdapter.ACTION_STATE_CHANGED);
            registerReceiver(mBluetoothReceiver, filter);
        }
    }

    protected boolean hasBluetoothPermission() {
        return hasBluetoothPermission;
    }

    protected boolean isBluetoothConnectionRequired() {
        return hasBluetoothPermission();
    }

    public ObservationKey getKey() {
        return key;
    }

    public AppSource getSource() {
        return source;
    }

    public void registerDevice(String name, Map<String, String> attributes) {
        registerDevice(null, name, attributes);
    }

    public void registerDevice(String sourceIdHint, String name, Map<String, String> attributes) {
        logger.info("Registering source {} with attributes {}", source, attributes);
        if (source.getSourceId() != null) {
            DeviceManager<T> localManager = getDeviceManager();
            if (localManager != null) {
                localManager.didRegister(source);
            }
            return;
        }
        source.setSourceName(name);
        source.setAttributes(attributes);
        // not yet registered
        if (source.getSourceId() == null) {
            if (ManagementPortalService.isEnabled()) {
                // do registration with management portal
                final Handler handler = new Handler(getMainLooper());
                ManagementPortalService.registerSource(this, source,
                        new ResultReceiver(handler) {
                    @Override
                    protected void onReceiveResult(int resultCode, Bundle result) {
                        updateRegistration(resultCode, result, handler);
                    }
                });
                return;
            } else {
                // self-register
                if (sourceIdHint == null) {
                    source.setSourceId(RadarConfiguration.getOrSetUUID(this, SOURCE_ID_KEY));
                } else {
                    source.setSourceId(sourceIdHint);
                }
                key.setSourceId(source.getSourceId());
            }
        }
        DeviceManager<T> localManager = getDeviceManager();
        if (localManager != null) {
            localManager.didRegister(source);
        }
    }

    private void updateRegistration(int resultCode, Bundle result, Handler handler) {
        AppSource updatedSource = null;
        if (resultCode == MANAGEMENT_PORTAL_REGISTRATION && result != null && result.containsKey(SOURCE_KEY)) {
            updatedSource = result.getParcelable(SOURCE_KEY);
        }
        if (updatedSource == null) {
            // try again in a minute
            handler.postDelayed(() -> stopDeviceManager(unsetDeviceManager()),
                    ThreadLocalRandom.current().nextLong(1_000L, 120_000L));
            return;
        }
        AppAuthState auth = AppAuthState.Builder.from(result).build();
        if (auth.isInvalidated()) {
            logger.info("New source ID requires new OAuth2 JWT token.");
            broadcastServerStatus(UNAUTHORIZED);
        }
        key.setProjectId(auth.getProjectId());
        key.setUserId(auth.getUserId());
        key.setSourceId(updatedSource.getSourceId());
        source.setSourceId(updatedSource.getSourceId());
        source.setSourceName(updatedSource.getSourceName());
        source.setExpectedSourceName(updatedSource.getExpectedSourceName());
        DeviceManager<T> localManager = getDeviceManager();
        if (localManager != null) {
            localManager.didRegister(source);
        }
    }

    @Override
    public String toString() {
        DeviceManager localManager = getDeviceManager();
        if (localManager == null) {
            return getClass().getSimpleName() + "<null>";
        } else {
            return getClass().getSimpleName() + "<" + localManager.getName() + ">";
        }
    }
}

File: radar-commons-android/src/main/java/org/radarbase/android/device/DeviceService.kt
/*
 * Copyright 2017 The Hyve
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.radarbase.android.device

import android.app.Service
import android.bluetooth.BluetoothAdapter
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.os.Bundle
import android.os.Process.THREAD_PRIORITY_BACKGROUND
import androidx.annotation.CallSuper
import androidx.localbroadcastmanager.content.LocalBroadcastManager
import org.apache.avro.specific.SpecificRecord
import org.radarbase.android.RadarConfiguration
import org.radarbase.android.auth.*
import org.radarbase.android.auth.portal.SourceType
import org.radarbase.android.data.DataHandler
import org.radarbase.android.device.DeviceServiceProvider.Companion.NEEDS_BLUETOOTH_KEY
import org.radarbase.android.radarApp
import org.radarbase.android.util.BundleSerialization
import org.radarbase.android.util.ManagedServiceConnection
import org.radarbase.android.util.SafeHandler
import org.radarbase.android.util.send
import org.radarcns.kafka.ObservationKey
import org.slf4j.LoggerFactory
import java.io.IOException
import java.util.HashSet
import kotlin.collections.HashMap

/**
 * A service that manages a DeviceManager and a TableDataHandler to send addToPreferences the data of a
 * wearable device and send it to a Kafka REST proxy.
 *
 * Specific wearables should extend this class.
 */
abstract class DeviceService<T : BaseDeviceState> : Service(), DeviceStatusListener, LoginListener {
    val key = ObservationKey()

    /** Stops the device when bluetooth is disabled.  */
    private val mBluetoothReceiver = object : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {
            val action = intent.action

            if (action == BluetoothAdapter.ACTION_STATE_CHANGED) {
                val state = intent.getIntExtra(
                        BluetoothAdapter.EXTRA_STATE, BluetoothAdapter.ERROR)
                when (state) {
                    BluetoothAdapter.STATE_TURNING_OFF, BluetoothAdapter.STATE_OFF -> {
                        logger.warn("Bluetooth is off")
                        handler.execute { stopDeviceManager(unsetDeviceManager()) }
                    }
                    else -> logger.debug("Bluetooth is in state {}", state)
                }
            }
        }
    }
    @get:Synchronized
    @set:Synchronized
    var dataHandler: DataHandler<ObservationKey, SpecificRecord>? = null
    @get:Synchronized
    var deviceManager: DeviceManager<T>? = null
        private set
    private lateinit var mBinder: DeviceBinder<T>
    private var hasBluetoothPermission: Boolean = false
    lateinit var sources: Set<SourceMetadata>
    lateinit var sourceTypes: Set<SourceType>
    private lateinit var authConnection: AuthServiceConnection
    protected lateinit var config: RadarConfiguration
    private lateinit var radarConnection: ManagedServiceConnection<org.radarbase.android.IRadarBinder>
    private lateinit var handler: SafeHandler
    private var startFuture: SafeHandler.HandlerFuture? = null
    private lateinit var broadcaster: LocalBroadcastManager

    val state: T
        get() {
            return deviceManager?.state ?: defaultState.apply {
                id.apply {
                    setProjectId(key.getProjectId())
                    setUserId(key.getUserId())
                    setSourceId(key.getSourceId())
                }
            }
        }

    /**
     * Default state when no device manager is active.
     */
    protected abstract val defaultState: T

    private val expectedSourceNames: Set<String>
        get() = HashSet<String>(sources
                .map { it.expectedSourceName }
                .filter { it != null })

    open val isBluetoothConnectionRequired: Boolean
        get() = hasBluetoothPermission()

    @CallSuper
    override fun onCreate() {
        logger.info("Creating DeviceService {}", this)
        super.onCreate()
        sources = HashSet()
        sourceTypes = HashSet()
        broadcaster = LocalBroadcastManager.getInstance(this)

        mBinder = createBinder()

        if (isBluetoothConnectionRequired) {
            val intentFilter = IntentFilter(BluetoothAdapter.ACTION_STATE_CHANGED)
            registerReceiver(mBluetoothReceiver, intentFilter)
        }

        authConnection = AuthServiceConnection(this, this)
        authConnection.bind()

        radarConnection = ManagedServiceConnection(this, radarApp.radarService)
        radarConnection.onBoundListeners.add { binder ->
            dataHandler = binder.dataHandler
            handler.execute {
                startFuture?.runNow()
            }
        }
        radarConnection.bind()
        handler = SafeHandler(javaClass.simpleName, THREAD_PRIORITY_BACKGROUND)

        config = radarApp.configuration

        deviceManager = null
        startFuture = null
    }

    @CallSuper
    override fun onDestroy() {
        logger.info("Destroying DeviceService {}", this)
        super.onDestroy()

        radarConnection.unbind()
        authConnection.unbind()

        if (isBluetoothConnectionRequired) {
            // Unregister broadcast listeners
            unregisterReceiver(mBluetoothReceiver)
        }
        handler.stop { stopDeviceManager(unsetDeviceManager()) }

        radarApp.onDeviceServiceDestroy(this)
    }

    @CallSuper
    protected open fun configure(configuration: RadarConfiguration) {
        val localManager = deviceManager
        if (localManager != null) {
            configureDeviceManager(localManager, configuration)
        }
    }

    protected open fun configureDeviceManager(manager: DeviceManager<T>, configuration: RadarConfiguration) {}

    override fun onBind(intent: Intent): DeviceBinder<T> {
        doBind(intent, true)
        return mBinder
    }

    @CallSuper
    override fun onRebind(intent: Intent) {
        doBind(intent, false)
    }

    private fun doBind(intent: Intent, firstBind: Boolean) {
        logger.debug("Received (re)bind in {}", this)
        val extras = BundleSerialization.getPersistentExtras(intent, this) ?: Bundle()
        onInvocation(extras)

        radarApp.onDeviceServiceInvocation(this, extras, firstBind)
    }

    override fun onUnbind(intent: Intent): Boolean {
        logger.debug("Received unbind in {}", this)
        return true
    }

    override fun deviceFailedToConnect(name: String) {
        broadcaster.send(DEVICE_CONNECT_FAILED) {
            putExtra(DEVICE_SERVICE_CLASS, javaClass.name)
            putExtra(DEVICE_STATUS_NAME, name)
        }
    }

    private fun broadcastDeviceStatus(name: String?, status: DeviceStatusListener.Status) {
        broadcaster.send(DEVICE_STATUS_CHANGED) {
            putExtra(DEVICE_STATUS_CHANGED, status.ordinal)
            putExtra(DEVICE_SERVICE_CLASS, javaClass.name)
            name?.let { putExtra(DEVICE_STATUS_NAME, it) }
        }
    }

    override fun deviceStatusUpdated(deviceManager: DeviceManager<*>, status: DeviceStatusListener.Status) {
        if (status == DeviceStatusListener.Status.DISCONNECTED) {
            handler.execute(true) {
                if (this.deviceManager === deviceManager) {
                    this.deviceManager = null
                }
                stopDeviceManager(deviceManager)
            }
        }
        broadcastDeviceStatus(deviceManager.name, status)
    }

    @Synchronized
    private fun unsetDeviceManager(): DeviceManager<*>? {
        val tmpManager = deviceManager
        deviceManager = null
        handler.stop { }
        return tmpManager
    }

    private fun stopDeviceManager(deviceManager: DeviceManager<*>?) {
        if (deviceManager != null) {
            if (!deviceManager.isClosed) {
                try {
                    deviceManager.close()
                } catch (e: IOException) {
                    logger.warn("Failed to close device scanner", e)
                }

            }
        }
    }

    /**
     * New device manager for the current device.
     */
    protected abstract fun createDeviceManager(): DeviceManager<T>

    fun startRecording(acceptableIds: Set<String>) {
        if (key.getUserId() == null) {
            throw IllegalStateException("Cannot start recording: user ID is not set.")
        }
        val expectedNames = expectedSourceNames
        val actualIds = if (expectedNames.isEmpty()) acceptableIds else expectedNames

        handler.start()
        handler.execute {
            val localManager = deviceManager
            if (localManager == null) {
                if (isBluetoothConnectionRequired && BluetoothAdapter.getDefaultAdapter() != null && !BluetoothAdapter.getDefaultAdapter().isEnabled) {
                    logger.error("Cannot start recording without Bluetooth")
                    return@execute
                }
                if (radarConnection.binder != null && dataHandler != null) {
                    logger.info("Starting recording")
                    if (deviceManager == null) {
                        createDeviceManager().also { manager ->
                            deviceManager = manager
                            configureDeviceManager(manager, config)
                            manager.start(actualIds)
                        }
                    }
                } else {
                    startFuture = handler.delay(100) {
                        startFuture?.let {
                            startFuture = null
                            startRecording(acceptableIds)
                        }
                    }
                }
            }
        }
    }

    fun stopRecording() {
        handler.execute {
            startFuture?.let {
                it.cancel()
                startFuture = null
            }
            stopDeviceManager(unsetDeviceManager())
            logger.info("Stopped recording {}", this)
        }
    }

    override fun loginSucceeded(manager: LoginManager?, authState: AppAuthState) {
        key.setProjectId(authState.projectId)
        key.setUserId(authState.userId)
    }

    override fun loginFailed(manager: LoginManager?, ex: Exception?) {

    }

    /**
     * Override this function to get any parameters from the given intent.
     * Bundle classloader needs to be set correctly for this to work.
     *
     * @param bundle intent extras that the activity provided.
     */
    @CallSuper
    fun onInvocation(bundle: Bundle) {
        setHasBluetoothPermission(bundle.getBoolean(NEEDS_BLUETOOTH_KEY, false))

        configure(config)
    }

    /** Get the service local binder.  */
    protected fun createBinder(): DeviceBinder<T> {
        return DeviceBinder(this)
    }

    protected fun setHasBluetoothPermission(isRequired: Boolean) {
        val oldBluetoothNeeded = isBluetoothConnectionRequired
        hasBluetoothPermission = isRequired
        val newBluetoothNeeded = isBluetoothConnectionRequired

        if (oldBluetoothNeeded && !newBluetoothNeeded) {
            unregisterReceiver(mBluetoothReceiver)
        } else if (!oldBluetoothNeeded && newBluetoothNeeded) {
            // Register for broadcasts on BluetoothAdapter state change
            val filter = IntentFilter(BluetoothAdapter.ACTION_STATE_CHANGED)
            registerReceiver(mBluetoothReceiver, filter)
        }
    }

    protected fun hasBluetoothPermission(): Boolean {
        return hasBluetoothPermission
    }

    private fun registerDevice(type: SourceType, name: String, attributes: Map<String, String>) {
        logger.info("Registering source {} with attributes {}", type, attributes)

        val source = SourceMetadata(type).apply {
            sourceName = name
            this.attributes = attributes
        }

        authConnection.binder?.registerSource(source,
                { authState, updatedSource ->
                    key.setProjectId(authState.projectId)
                    key.setUserId(authState.userId)
                    key.setSourceId(updatedSource.sourceId)
                    source.sourceId = updatedSource.sourceId
                    source.sourceName = updatedSource.sourceName
                    source.expectedSourceName = updatedSource.expectedSourceName
                    deviceManager?.didRegister(source)
                },
                { handler.delay(300_000L) { registerDevice(type, name, attributes) } })
                ?: handler.delay(300_000L) { registerDevice(type, name, attributes) }
    }

    open fun ensureRegistration(id: String?, name: String, attributes: Map<String, String>): Boolean {
        val fullAttributes = HashMap(attributes).apply {
            put("physicalId", id ?: "")
            put("physicalName", name)
        }
        return if (sources.isEmpty()) {
            matchingSourceType?.let { registerDevice(it, name, fullAttributes) } != null
        } else {
            val matchingSource = sources
                    .find { source ->
                        if (source.matches(id, name)) {
                            true
                        } else if (id != null && source.attributes["physicalId"]?.isEmpty() == false) {
                            source.attributes["physicalId"] == id
                        } else if (source.attributes["physicalName"]?.isEmpty() == false) {
                            source.attributes["physicalName"] == name
                        } else {
                            false
                        }
                    }

            if (matchingSource == null) {
                matchingSourceType?.let { registerDevice(it, name, fullAttributes) } != null
            } else {
                deviceManager?.didRegister(matchingSource)
                true
            }
        }
    }

    private val matchingSourceType: SourceType?
        get() = sourceTypes
                    .filter { it.hasDynamicRegistration }
                    .sortedBy { it.catalogVersion }
                    .firstOrNull()

    override fun toString(): String {
        val localManager = deviceManager
        return if (localManager == null) {
            javaClass.simpleName + "<null>"
        } else {
            javaClass.simpleName + "<" + localManager.name + ">"
        }
    }

    companion object {
        private const val PREFIX = "org.radarcns.android."
        const val SERVER_STATUS_CHANGED = PREFIX + "ServerStatusListener.Status"
        const val SERVER_RECORDS_SENT_TOPIC = PREFIX + "ServerStatusListener.topic"
        const val SERVER_RECORDS_SENT_NUMBER = PREFIX + "ServerStatusListener.lastNumberOfRecordsSent"
        const val CACHE_TOPIC = PREFIX + "DataCache.topic"
        const val CACHE_RECORDS_UNSENT_NUMBER = PREFIX + "DataCache.numberOfRecords.first"
        const val DEVICE_SERVICE_CLASS = PREFIX + "DeviceService.getClass"
        const val DEVICE_STATUS_CHANGED = PREFIX + "DeviceStatusListener.Status"
        const val DEVICE_STATUS_NAME = PREFIX + "DeviceManager.getName"
        const val DEVICE_CONNECT_FAILED = PREFIX + "DeviceStatusListener.deviceFailedToConnect"

        private val logger = LoggerFactory.getLogger(DeviceService::class.java)
    }
}
